### PR TITLE
[MkFit] Fix UBSAN error - skip processing of layers with no hits.

### LINKED ISSUE #37136: [UBSAN] MkFinder is assigning `nullptr` to a reference
Workflow 10004.0 for the UBSAN IB reports 
```
/data/cmsbld/jenkins/workspace/build-any-ib/w/slc7_amd64_gcc11/external/gcc/11.2.1-f478fee2760dbd22aaabb4e3a8fe1640/include/c++/11.2.1/bits/stl_vector.h:1046:34: runtime error: reference binding to null pointer of type 'value_type'
    #0 0x2b5fbc317a13  (/cvmfs/cms-ib.cern.ch/nweek-02722/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-02-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCore.so+0x730a13)
    #1 0x2b5fbc33116b  (/cvmfs/cms-ib.cern.ch/nweek-02722/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-02-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCore.so+0x74a16b)
    #2 0x2b5fbbbbaefc  (/cvmfs/cms-ib.cern.ch/nweek-02722/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-02-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCMS.so+0x50efc)
    #3 0x2b5fbb6d0d8e  (/cvmfs/cms-ib.cern.ch/nweek-02722/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-02-1100/lib/slc7_amd64_gcc11/pluginRecoTrackerMkFitPlugins.so+0x65cd8e)
```

### PR DESCRIPTION
**Pull request by osschar**
When processing layers with no hits the LayerOfHits vector was in default state with storage pointer set to nullptr.

Loaders of hits for vector processing were still being initialized, resulting in doing pointer arithmetic with nullptr -- though they were never actually used.

This should fix #37136. Note that the wf there was 10004.0, single photon 10GeV and there were indeed several occurrences of empty mkfit hit containers.

### CODE REVIEW

**Comment by makortel** [comment]
test parameters:
  - release = CMSSW_12_3_UBSAN_X
  - workflow = 10004.0

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
@smuzaffar this has been running tests for 3 days (!?).
Is there something stuck?

**Comment by smuzaffar** [comment]
please test

**Comment by clacaputo** [comment]
> @mmusich , UBSAN IB comparison does not work. The job timedout after 16 hours ( https://cmssdt.cern.ch/jenkins/job/compare-root-files-short-matrix/48727/ ) . Check the relvals for wf 10004.0 to see if USBAN issue is fixed or not?
> 
> I have restarted the comparison job now but I am afraid it will timeout again.

Hi @smuzaffar, where can I check wf 10004.0 relvals for UBSAN?

**Comment by smuzaffar** [comment]
URL (i.e. https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdca02/22856)  for `cms/37146/UBSAN/slc7_amd64_gcc11` context points to UBSAN results. Go there and then click on the `summary.html` page to see the UBSAN results

**Comment by smuzaffar** [comment]
ok, I have forced set the status for `ubsan` comparison check, but should add the comment for ubsan now

**Comment by clacaputo** [comment]
If I check the output in `10004.0` step3 log, the runtime error is still there

```
/cvmfs/cms-ib.cern.ch/nweek-02722/slc7_amd64_gcc11/external/gcc/11.2.1-f478fee2760dbd22aaabb4e3a8fe1640/include/c++/11.2.1/bits/stl_vector.h:1046:34: runtime error: reference binding to null pointer of type 'value_type'
    #0 0x2abc3b477a03  (/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37146/22856/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCore.so+0x730a03)
    #1 0x2abc3b49126b  (/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37146/22856/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCore.so+0x74a26b)
    #2 0x2abc3ad1aefc  (/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCMS.so+0x50efc)
    #3 0x2abc3a82fd8e  (/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/pluginRecoTrackerMkFitPlugins.so+0x65cd8e)
```

**Comment by mmusich** [comment]
> If I check the output in 10004.0 step3 log, the runtime error is still there

I observe the same locally:
```console
cmsrel CMSSW_12_3_UBSAN_X_2022-03-04-1100
cd CMSSW_12_3_UBSAN_X_2022-03-04-1100/src/
cmsenv
git cms-merge-topic 37146
scramv1 b -j 20
runTheMatrix.py -l 10004.0 -t 4 -j 8
more 10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log | grep libRecoTrackerMkFitCore | wc -l 
2
```

**Comment by osschar** [comment]
Oh rats. How does one setup the build environment for this?
Reading gcc man page ... is USER_CXXFLAGS="-fsanitize=undefined -fsanitize=null" sufficient?

**Comment by osschar** [comment]
There were also cases where MkFitProducer input seed vector was empty. The commit above bypasses all calls to mkFit and registers an empty track vector into Event.

This helped get to the error in gdb:
```
USER_CXXFLAGS="-fsanitize=undefined -fsanitize=null -g -O0" scram b -j16
gdb --args cmsRun ....
### wait for stuff to get loaded
b __ubsan::ScopedReport::~ScopedReport
### there are multiple functions in namespace __ubsan one can break on
```

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
with this version of the code and the recipe at https://github.com/cms-sw/cmssw/pull/37146#issuecomment-1061736202 I get:

```console
$ more 10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log | grep libRecoTrackerMkFitCore | wc -l 
0
```
so we should be in the clear.

**Comment by clacaputo** [comment]
I'm sees some reco differences in wf `139.001` in `recoTracks_generalTracks__reRECO_obj`
For example
all_OldVSNew_RunMinimumBias2021wf139p001c_recoTracks_generalTracks__reRECO_obj_p

Although it's infinitesimal, I wouldn't expect it considering the purely technical nature of this PR. Can be related to the wf being a MinimumBias wf?

File: RecoTracker/MkFitCore/src/MkBuilder.cc [hunk 1]
```diff
@@ -800,7 +799,9 @@ namespace mkfit {
                                                          layer_plan_it.is_pickup_only(),
                                                          iteration_dir);
 
-          if (layer_plan_it.is_pickup_only() || theEndCand == 0)
+          dprintf("  Number of candidates to process: %d, nHits in layer: %d\n", theEndCand, layer_of_hits.nHits());
+
+          if (layer_plan_it.is_pickup_only() || theEndCand == 0 || layer_of_hits.nHits() == 0)
```
Comment (slava77) [inline RecoTracker/MkFitCore/src/MkBuilder.cc:804]: is the skip here on empty layer completely benign? Would we appropriately get the invalid/missing hits and material updates processed?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by clacaputo** [comment]
+reconstruction

- technical PR, fix for UBSAN
- checked locally
```bash
cmsrel CMSSW_12_3_UBSAN_X_2022-03-04-1100
cd CMSSW_12_3_UBSAN_X_2022-03-04-1100/src/
cmsenv
git cms-merge-topic 37146
scramv1 b -j 20
runTheMatrix.py -l 10004.0 -t 4 -j 8
more 10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log | grep libRecoTrackerMkFitCore | wc -l 
0
```
- no reco changes expected or observed after b548ebf4d8d5bc55dc06c3444718bd0d271588fc

**Comment by perrotta** [comment]
+1

**Comment by slava77** [comment]
@cms-sw/core-l2 @perrotta @qliphy 
do you need a backport to 12_3_X?

**Comment by makortel** [comment]
I suppose it would be useful (to minimize UB in 12_3).

**Pull request merged**